### DISCUSSION TITLE
Proper way to use jotai/query with parameters

### DISCUSSION THREAD

**Discussion by TeaBough** [pmndrs/jotai#1505] [Q&A] [answered]
Hi ! I'm having a hard time figuring out the best way to handle a simple case of fetching a single element.
My model is this one : I have Folders that contains Files. I'd like to display a page displaying a single File contained in a folder.
The url coud be something like that : `.../folders/12345/files/1`
Previously, with only react-query, I used this : 

```
export default function useFile(fileId?: string) {
  const { folter } = useCurrentFolder();
  return useQuery<File, AxiosError>(
    [folder.id, 'FOLDER_KEY', fileId],
    () =>
      fetchFile({
        folderId: folder.id,
        fileId,
      }),
  );
}
```

Now with jotai-query I'm doing this : 

```
export function useFile(fileId: string) {
    const { folder } = useCurrentFolder();

    const [fileAtom] = useMemo(
        () =>
            atomsWithQuery(() => ({
                queryKey: [
                    folder.id,
                    'FOLDER_KEY',
                    fileId,
                ],
                queryFn: async () => {
                    return fetchFile({
                        folderId: folder.id,
                        fileId,
                    });
                },
            })),
        [folder.id, fileId],
    );
    const [data] = useAtom(todoItemAtom);
    return data;
}
```

Is this the correct way to do that ?

**Comment by dai-shi** [comment]
That's unfortunate but correct approach.
You want to create atoms outside Suspense boundary. (hm, but maybe query cache avoids infinite loop.)

The better is if you could somehow create `currentFolderIdAtom` and `fileIdAtom`.
But unless the library providing `useCurrentFolder` provides "vanilla" API, it'd be difficult.

**Comment by TeaBough** [comment, marked as answer]
`useCurrentFolder` is just an utility hook to get the current folder value that is in the global context of the app.
But suppose I don't use `useCurrentFolder`, I still need a way to put the `fileId` in an `fileIdAtom`.
So I guess I should create a `fileIdAtom` outside `Suspense` and somewhere, where my route is matching do something like that :
```
const fileIdAtom = atom()
const setFileIdAtom = atom(
    null,
    (get, set, fileId) => {
        set(fileIdAtom, fileId)
    })


export default function FileRoot(props) {
    // the router is giving me this props
    const { fileId } = props;
    const [, setFileId] = useAtom(setFileIdAtom)
    useEffect(() => {
        setFileId(fileId)
    }, [fileId])
}
```

**Reply by dai-shi** [reply]
Yeah, that's the correct approach.
If we could directly access to the router, it might be possible to bind `fileId`, but that's not very nice DX-wise.

This is a general problem of binding props and atoms. So far, `useEffect` is a right solution, but we hope to find some other solutions...

**Reply by TeaBough** [reply]
👍 Thank you @dai-shi for those quick answers. It's much clearer to me now how to fit jotai with react-query 🎉
